### PR TITLE
Expose printing functionality of emitter as a public API

### LINKED ISSUE #13762: Expose pretty printer as public API
Expose the pretty printer functionality of the emitter as a public API.

### PR DESCRIPTION
**Pull request by rbuckton**
This exposes a new public `createPrinter` export on the `ts` namespace that can be used to print a TypeScript AST as-is (without transformation):

```ts
// public API
declare namespace ts {
  // subset of CompilerOptions
  interface PrinterOptions {
    target?: ScriptTarget;
    removeComments?: boolean;
    newLine?: NewLineKind;
  }

  // custom hooks for name generation, emit notification, and substitution
  interface PrintHandlers {
    hasGlobalName?: (name: string) => boolean;
    onEmitNode?: (hint: EmitHint, node: Node, emitCallback: (hint: EmitHint, node: Node) => void) => void;
    onSubstituteNode?: (hint: EmitHint, node: Node, emitCallback: (hint: EmitHint, node: Node) => void) => void;
  }

  // printer service
  interface Printer {
    printNode(hint: EmitHint, node: Node, sourceFile: SourceFile): string;
    printFile(sourceFile: SourceFile): string;
    printBundle(bundle: Bundle): string;
  }
  
  // printer factory
  function createPrinter(printerOptions?: PrinterOptions, handlers?: PrintHandlers): Printer;

  // bundles output (for --outFile)
  interface Bundle extends Node {
    kind: SyntaxKind.Bundle;
    sourceFiles: SourceFile[];
  }
}
```

To support our internal emit needs, we also have the following internal API for printing:

```ts
// internal API
declare namespace ts {
  // internal hooks to support source map and helper emit
  interface PrintHandlers {
    /*@internal*/ onEmitSourceMapOfNode?: (hint: EmitHint, node: Node, emitCallback: (hint: EmitHint, node: Node) => void) => void;
    /*@internal*/ onEmitSourceMapOfToken?: (node: Node, token: SyntaxKind, pos: number, emitCallback: (token: SyntaxKind, pos: number) => number) => number;
    /*@internal*/ onEmitSourceMapOfPosition?: (pos: number) => void;
    /*@internal*/ onEmitHelpers?: (node: Node, writeLines: (text: string) => void) => void;
    /*@internal*/ onSetSourceFile?: (node: SourceFile) => void;
  }

  // subset of compiler options needed to support source map emit
  export interface PrinterOptions {
    /*@internal*/ sourceMap?: boolean;
    /*@internal*/ inlineSourceMap?: boolean;
    /*@internal*/ extendedDiagnostics?: boolean;
  }

  // print a node, file, or bundle to an existing EmitTextWriter
  export interface Printer {
    /*@internal*/ writeNode(hint: EmitHint, node: Node, sourceFile: SourceFile, writer: EmitTextWriter): void;
    /*@internal*/ writeFile(sourceFile: SourceFile, writer: EmitTextWriter): void;
    /*@internal*/ writeBundle(bundle: Bundle, writer: EmitTextWriter): void;
  }  
}
```

The public API for printing *does not* support source maps currently as this is heavily tied into the emitter. We may choose to support this via a later PR.

This further extends the work started in #5595 with the overall goal of making significant portions of the new transformation and emit pipeline part of the public API. This is also intended to help support ongoing effort into refactoring and code actions.

Fixes #13762

### CODE REVIEW

File: src/compiler/comments.ts [hunk 1]
```diff
@@ -198,9 +198,9 @@ namespace ts {
             }
 
             // Leading comments are emitted at /*leading comment1 */space/*leading comment*/space
-            emitPos(commentPos);
+            if (emitPos) emitPos(commentPos);
```
Comment (vladima) [inline src/compiler/comments.ts]: why this check is necessary? If `emitPos` can be undefined, can you please reflect this in its type?

**Comment by rbuckton** [comment]
@mhegazy The tasks involved in making the API public can be found [here](https://github.com/Microsoft/TypeScript/projects/4).

File: src/compiler/comments.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/comments.ts]: Source map support is not currently exposed publicly, so a printer created through the public API would not be able to provide an `emitPos` callback. I'll amend the parameter to include `undefined`.

**Comment by rbuckton** [comment]
@sandersn both (1) and (2) are incorrect, there were no changes to that functionality of emitter, other than moving `createPrinter` out of `emitFiles`. Github is just showing a significant amount of changes due to some minor reorganization as most of the existing functions were moved down and `emitHelpers` was moved up as they are now in different closures.

**Comment by aleksey-bykov** [comment]
nice

**Pull request merged**